### ISSUE TITLE
Smart Disable-Wait functions based on events

### ISSUE THREAD

**Issue by RafaBenard** [slgobinath/SafeEyes#261]
Hi there,

I've been using this software for a while and it is pretty useful. However one thing that becomes annoying if you're writting a lot is the fact that the Alert can pop-up in the middle of a conversation/idea despite the keyboard being activated.

So, one idea is to have sort of a Smart Pause on the break screens. For example, something that will not trigger this window when you're typing (i.e.: Check 15s before the alert is going to appear) and automatically pause the alert until 15s have passed.

Thoughts?

**Comment by slgobinath** [comment]
Hi,
The existing `Do not disturb plugin` should do this for you. Do you have `xprintidle` installed and `Do not disturb plugin` enabled?

If you already have them, please quit currently running safe eyes instance and start it from a terminal in debug mode `safeeyes --debug`. Use it for a while and share the `~/safeeyes.log` file here.

**Comment by RafaBenard** [comment]
Hi @slgobinath 

Yes! I've got `xprintidle` installed and the `Do not disturb plugin` on as well.

Here's the output of the `~/safeeyes.log` file:
```2018-05-24 14:28:43,740 [INFO]:[MainThread] Starting Safe Eyes
2018-05-24 14:28:43,792 [INFO]:[MainThread] Initialize the break screen
2018-05-24 14:28:43,792 [INFO]:[MainThread] Load all the plugins
2018-05-24 14:28:43,792 [INFO]:[MainThread] Initialize the core
2018-05-24 14:28:43,793 [INFO]:[MainThread] Successfully loaded <module 'donotdisturb.plugin' from '/usr/lib/python3/dist-packages/safeeyes/plugins/donotdisturb/plugin.py'>
2018-05-24 14:28:43,795 [INFO]:[MainThread] Successfully loaded <module 'notification.plugin' from '/usr/lib/python3/dist-packages/safeeyes/plugins/notification/plugin.py'>
2018-05-24 14:28:43,795 [INFO]:[MainThread] Successfully loaded <module 'audiblealert.plugin' from '/usr/lib/python3/dist-packages/safeeyes/plugins/audiblealert/plugin.py'>
2018-05-24 14:28:43,797 [INFO]:[MainThread] Successfully loaded <module 'trayicon.plugin' from '/usr/lib/python3/dist-packages/safeeyes/plugins/trayicon/plugin.py'>
2018-05-24 14:28:43,797 [INFO]:[MainThread] Successfully loaded <module 'smartpause.plugin' from '/usr/lib/python3/dist-packages/safeeyes/plugins/smartpause/plugin.py'>
2018-05-24 14:28:43,798 [INFO]:[MainThread] Successfully loaded <module 'screensaver.plugin' from '/usr/lib/python3/dist-packages/safeeyes/plugins/screensaver/plugin.py'>
2018-05-24 14:28:43,799 [INFO]:[MainThread] Successfully loaded <module 'healthstats.plugin' from '/usr/lib/python3/dist-packages/safeeyes/plugins/healthstats/plugin.py'>
2018-05-24 14:28:43,799 [DEBUG]:[MainThread] Initialize Skip Fullscreen plugin
2018-05-24 14:28:43,799 [DEBUG]:[MainThread] Initialize Notification plugin
2018-05-24 14:28:43,799 [DEBUG]:[MainThread] Initialize Audible Alert plugin
2018-05-24 14:28:43,799 [DEBUG]:[MainThread] Initialize Tray Icon plugin
2018-05-24 14:28:43,804 [DEBUG]:[MainThread] Initialize Smart Pause plugin
2018-05-24 14:28:43,804 [DEBUG]:[MainThread] Initialize Screensaver plugin
2018-05-24 14:28:43,805 [DEBUG]:[MainThread] Initialize Health Stats plugin
2018-05-24 14:28:43,805 [INFO]:[MainThread] Setting up an RPC server on port 7200
2018-05-24 14:28:43,805 [INFO]:[MainThread] Start the RPC server
2018-05-24 14:28:43,806 [DEBUG]:[MainThread] Start Smart Pause plugin
2018-05-24 14:28:43,806 [INFO]:[MainThread] Start Safe Eyes core
2018-05-24 14:28:43,806 [INFO]:[Thread-3] Waiting for 20 minutes until next break
2018-05-24 14:28:43,808 [INFO]:[MainThread] Update next break information
2018-05-24 14:48:43,807 [INFO]:[Thread-3] Pre-break waiting is over
2018-05-24 14:48:43,809 [INFO]:[MainThread] Show the notification
2018-05-24 14:48:43,822 [INFO]:[Thread-5] Wait for 10 seconds before the break
2018-05-24 14:48:53,824 [INFO]:[MainThread] Searching for full-screen application
2018-05-24 14:48:53,840 [INFO]:[MainThread] Close pre-break notification
2018-05-24 14:48:53,845 [INFO]:[Thread-7] Lock the keyboard
2018-05-24 14:48:53,847 [INFO]:[MainThread] Show break screens in 2 display(s)
2018-05-24 14:48:53,915 [INFO]:[MainThread] Moved break screen to Display[1920, 0]
2018-05-24 14:48:53,919 [INFO]:[MainThread] Moved break screen to Display[0, 0]
2018-05-24 14:49:08,864 [INFO]:[MainThread] Break is terminated automatically
2018-05-24 14:49:08,865 [INFO]:[MainThread] Close the break screen(s)
2018-05-24 14:49:08,865 [INFO]:[MainThread] Unlock the keyboard
2018-05-24 14:49:08,866 [INFO]:[MainThread] Playing audible alert
2018-05-24 14:49:08,874 [INFO]:[Thread-8] Waiting for 20 minutes until next break
2018-05-24 14:49:08,878 [INFO]:[MainThread] Update next break information
2018-05-24 14:49:19,770 [INFO]:[MainThread] Quit Safe Eyes
2018-05-24 14:49:19,771 [INFO]:[MainThread] Disable Safe Eyes
2018-05-24 14:49:19,772 [DEBUG]:[MainThread] Stop Smart Pause plugin
2018-05-24 14:49:19,773 [INFO]:[MainThread] Stop Safe Eye core
2018-05-24 14:49:19,774 [DEBUG]:[MainThread] Stop Notification plugin
2018-05-24 14:49:19,774 [INFO]:[Thread-8] Pre-break waiting is over
2018-05-24 14:49:19,775 [INFO]:[MainThread] Stop the RPC server```

I can see this:

```
2018-05-24 14:48:43,809 [INFO]:[MainThread] Show the notification
2018-05-24 14:48:43,822 [INFO]:[Thread-5] Wait for 10 seconds before the break
2018-05-24 14:48:53,824 [INFO]:[MainThread] Searching for full-screen application
2018-05-24 14:48:53,840 [INFO]:[MainThread] Close pre-break notification
2018-05-24 14:48:53,845 [INFO]:[Thread-7] Lock the keyboard
```

So, basicaly, the notification appears and then it waits 10s before the break, it searchs for the full-screen apps only but it doesn't consider the keyboard/mouse activities which is the recommendation.

**Comment by slgobinath** [comment]
I misunderstood your issue.
I think I got your point now. You mean at the exact time you get a break, if your keyboard is active then you need to skip the break. Am I right?

**Comment by slgobinath** [comment]
If that is the case, sorry I am against to this because I will never get a break. Instead, I have an alternative solution to skip break based on your current window.
Say you are writing a code in VSCode and you do not want a break while you are using VSCode. You can skip breaks based on the active window.

1. Open your editor and a terminal
2. Enter `xprop WM_CLASS` in terminal, hit enter and click on the editor window
3. It will print the window class of that app (For example I got `WM_CLASS(STRING) = "Navigator", "Firefox"` for Firefox and `WM_CLASS(STRING) = "code", "Code"` for VSCode)
3. Add the window class to never interrupt in  `Do Not Disturb` preference. (You can add multiple windo classes separated by comma)
![safeeyes_dnd](https://user-images.githubusercontent.com/4003168/39922630-f4378824-54ed-11e8-9164-c287f90b7b92.png)

I hope it may help you. If not please let me know.

**Comment by RafaBenard** [comment]
Hi,

Thanks for the response.

I'm sorry but that does not solve the issue. The request is to disable or actually delay the safe eyes popup based on certain events, not in windows or windows status'es.

For example, if you're typing something the alerts pops up and basically you're left with the dark background alert in the middle of typing a sentence, e-mail, search, game, etc. 

Basically, what I'm saying is that the alert becomes intrusive when you're using it on a working environment where you've to finish a sentence, consolidate an idea typing, etc. And in that we just end up skipping it indefinitely.

The proposal method for fixing is to have sort of a delay of the popup when you're typing, i.e. a plugin that will set a slight delay of 15s when you're typing, with an additional reminder perhaps so you know when to stop writing and you can also snooze a bit more the break.

Thanks!

**Comment by slgobinath** [comment]
I understood. It's a great idea and worth a new plugin.

**Comment by hartwork** [comment]
Getting right into my typing is the only pressing issue with SafeEyes for me right now. Would love to see that addressed.

**Comment by unode** [comment]
In #274 I got to pretty much the same solution as described here.

A mockup of how a break could be delayed is included in the other ticket and reads like:
```
00:00 - Short break is due for 15 seconds from now
00:05 - Notification is shown: `Short break will happen in 10 seconds or the nearest inactivity period`
00:14 - Keyboard activity detected delaying break 2 seconds
00:15 - Keyboard activity detected delaying break 2 seconds
00:16 - Keyboard activity detected delaying break 2 seconds
00:17 - Keyboard activity detected delaying break 2 seconds
(keyboard activity stops)
00:19 - Triggering break
```

So pretty much the same as described above.

**Comment by slgobinath** [comment]
Please note that this feature is added to the Smart Pause plugin. You can test it now by running Safe Eyes from source.
```
git clone https://github.com/slgobinath/SafeEyes.git
cd SafeEyes
python3 -m safeeyes --debug
```

Once Safe Eyes is running, goto Settings --> Plugins --> Smart Pause and enable "Postpone the next break until the system becomes idle"

I will release the new version after testing this feature for some days.

**Comment by slgobinath** [comment]
Included in 2.0.8
